girder_worker_utils/tests/tee_test.py

```diff
@@ -0,0 +1,177 @@
+    with TeeStdOut(pass_through=False):
+        print('Test String')
+
+    out, err = capfd.readouterr()
+    assert out == ''
+
+
+def test_tee_sys_write_stderr(capfd):
+    with TeeStdErr():
+        sys.stderr.write('Test String')
+        sys.stderr.flush()
+
+    out, err = capfd.readouterr()
+    assert err == 'Test String'
+
+
+def test_tee_stderr_sys_write_pass_through_false(capfd):
+    with TeeStdErr(pass_through=False):
+        sys.stderr.write('Test String')
+        sys.stderr.flush()
+
+    out, err = capfd.readouterr()
+    assert err == ''
+
+
+def test_tee_overwrites_write(capfd):
+    with TeeCapture() as o:
+        print('Test String')
+        assert o.buf == 'Test String\n'
+
+    out, err = capfd.readouterr()
+    assert out == 'Test String\n'
+
+
+def test_tee_overwrites_write_pass_through_false(capfd):
+    with TeeCapture(pass_through=False) as o:
+        print('Test String')
+        assert o.buf == 'Test String\n'
+
+    out, err = capfd.readouterr()
+    assert out == ''
+
+
+def test_tee_reset_function_resets(capfd):
+    original_stdout = sys.stdout
+
+    o = TeeCapture()
+
+    print('Test String')
+    out, err = capfd.readouterr()
+
+    assert o.buf == 'Test String\n'
+    assert out == 'Test String\n'
+
+    o.reset()
+
+    assert sys.stdout == original_stdout
+
+
+def test_tee_multiple_tee_objects_downstream():
+    original_stdout = sys.stdout
+    o1, o2, o3 = TeeCapture(), TeeCapture(), TeeCapture()
+
+    assert sys.stdout == o3
+    assert sys.stdout._downstream == o2
+    assert sys.stdout._downstream._downstream == o1
+    assert sys.stdout._downstream._downstream._downstream == original_stdout
+
+
+def test_tee_multiple_tee_objects_reset_o1():
+    original_stdout = sys.stdout
+    o1, o2, o3 = TeeCapture(), TeeCapture(), TeeCapture()
+
+    o1.reset()
+
+    assert o3._downstream == o2
+    assert o2._downstream == original_stdout
+
+
+def test_tee_multiple_tee_objects_reset_o2():
+    original_stdout = sys.stdout
+    o1, o2, o3 = TeeCapture(), TeeCapture(), TeeCapture()
+
+    o2.reset()
+
+    assert o3._downstream == o1
+    assert o1._downstream == original_stdout
+
+
+def test_tee_multiple_tee_objects_reset_o3():
+    original_stdout = sys.stdout
+    o1, o2, o3 = TeeCapture(), TeeCapture(), TeeCapture()
+
+    o3.reset()
+
+    assert sys.stdout == o2
+    assert o2._downstream == o1
+    assert o1._downstream == original_stdout
+
+
+def test_tee_multiple_tee_objects_reset_o3_then_o1():
+    original_stdout = sys.stdout
+    o1, o2, o3 = TeeCapture(), TeeCapture(), TeeCapture()
+
+    o3.reset()
+
+    assert sys.stdout == o2
+    assert o2._downstream == o1
+    assert o1._downstream == original_stdout
+
+    o1.reset()
+
+    assert sys.stdout == o2
+    assert o2._downstream == original_stdout
+
+
+def test_tee_multiple_tee_objects_reset_o3_then_o1_then_o2():
+    original_stdout = sys.stdout
+    o1, o2, o3 = TeeCapture(), TeeCapture(), TeeCapture()
+
```

Comment:
Add sanity check:

```
assert sys.stdout == o3
assert o3._downstream == o2
assert o2._downstream == o1
assert o1._downstream == original_stdout
```

Reply:
Also could you `print('foo')` and make sure each capture gets just one 'foo' and use `capsys` to verify the original stdout also just gets one 'foo'.

Reply:
So as not to muddle the ```test_tee_multiple_tee_objects_reset_o3_then_o1_then_o2``` I've written a separate test that runs earlier to handle this functionality.